iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -57,6 +47,78 @@ internal void SetRetryPolicy(IIotHubClientRetryPolicy retryPolicy)
                 Logging.Associate(this, _internalRetryHandler, nameof(SetRetryPolicy));
         }
 
+        public override async Task OpenAsync(CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, cancellationToken, nameof(OpenAsync));
+
+            try
+            {
+                await _internalRetryHandler
+                    .RunWithRetryAsync(
+                        async () =>
+                        {
+                            if (Logging.IsEnabled)
+                                Logging.Enter(this, cancellationToken, nameof(OpenAsync));
+
+                            try
+                            {
+                                // Will throw on error.
+                                await base.OpenAsync(cancellationToken).ConfigureAwait(false);
+
+                                if (_recoverSubscriptions)
+                                {
+
+                                    if (Logging.IsEnabled)
+                                        Logging.Info(this, "Attempting to recover subscriptions.", nameof(OpenAsync));
+
+                                    var tasks = new List<Task>(3);
+
+                                    // This is to ensure that, if previously enabled, the callback to receive direct methods is recovered.
+                                    if (_methodsEnabled)
+                                    {
+                                        tasks.Add(base.EnableMethodsAsync(cancellationToken));
+                                    }
+
+                                    // This is to ensure that, if previously enabled, the callback to receive twin properties is recovered.
+                                    if (_twinEnabled)
+                                    {
+                                        tasks.Add(base.EnableTwinPatchAsync(cancellationToken));
+                                    }
+
+                                    // This is to ensure that, if previously enabled, the callback to receive C2D messages is recovered.
+                                    if (_deviceReceiveMessageEnabled)
+                                    {
+                                        tasks.Add(base.EnableReceiveMessageAsync(cancellationToken));
+                                    }
+
+                                    if (tasks.Any())
+                                    {
+                                        await Task.WhenAll(tasks).ConfigureAwait(false);
+                                    }
+
+                                    if (Logging.IsEnabled)
+                                        Logging.Info(this, "Subscriptions recovered.", nameof(OpenAsync));
+                                }
+
+                                _recoverSubscriptions = true;
```

Comment:
Why is this set to true, if already in an if block confirming it was already true?

Reply:
It is set to true after the if block. This is to signal the method that if we have ever successfully opened the connection, we will need to recover the subscriptions with any future reconnection attempt.